### PR TITLE
[SPARK-28247][SS][TEST]Fix flaky test "query without test harness" on ContinuousSuite

### PR DESCRIPTION
**Pull request by zsxwing**
### What changes were proposed in this pull request?

This is another attempt to fix the flaky test "query without test harness" on ContinuousSuite.

`query without test harness` is flaky because it starts a continuous query with two partitions but assumes they will run at the same speed.

In this test, 0 and 2 will be written to partition 0, 1 and 3 will be written to partition 1. It assumes when we see 3, 2 should be written to the memory sink. But this is not guaranteed. We can add `if (currentValue == 2) Thread.sleep(5000)` at this line https://github.com/apache/spark/blob/b2a2b5d8206b7c09b180b8b6363f73c6c3fdb1d8/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala#L135 to reproduce the failure: `Result set Set([0], [1], [3]) are not a superset of Set(0, 1, 2, 3)!`

The fix is changing `waitForRateSourceCommittedValue` to wait until all partitions reach the desired values before stopping the query.

### Why are the changes needed?

Fix a flaky test.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests. Manually verify the reproduction I mentioned above doesn't fail after this change.

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @jose-torres

**Comment by viirya** [comment]
It seems to fail Scala 2.13 build.

> [error] /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala:63:53: type mismatch;
[error]  found   : scala.collection.MapView[Int,Long]
[error]  required: Map[Int,Long]
[error]             o.partitionToValueAndRunTimeMs.mapValues(_.value)
[error]                                                     ^

**Review by viirya — COMMENTED** [review]
looks reasonable.

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR FYI

**Review by HeartSaVioR — APPROVED** [review]
LGTM, thanks for the fix!

**Comment by HeartSaVioR** [comment]
Would we like to wait for @jose-torres to do the final review (and probably sign-off), or OK to go merging?

**Review by dongjoon-hyun — APPROVED** [review]

**Review by jose-torres — APPROVED** [review]
LGTM

**Comment by HeartSaVioR** [comment]
OK thanks everyone for reviewing. I'm going to merge this.

**Comment by HeartSaVioR** [comment]
Thanks @zsxwing for the fix! I merged this in master/3.1/3.0. I skipped 2.4 as it's unlikely that we'll want to maintain 2.4 version line further.

**Pull request closed without merging**
